isis/src/control/objs/BundleConstraint/BundleLidarRangeConstraint.cpp

```diff
@@ -0,0 +1,440 @@
+
+    // compute and store dX, dY, dZ between body-fixed coordinates of spacecraft and lidar point
+    m_dX = camPositionBodyFixed[0] - pointBodyFixed[0];
+    m_dY = camPositionBodyFixed[1] - pointBodyFixed[1];
+    m_dZ = camPositionBodyFixed[2] - pointBodyFixed[2];
+
+    m_rangeComputed = sqrt(m_dX*m_dX+m_dY*m_dY+m_dZ*m_dZ);
+
+    if (m_rangeComputed <= 0.0) {
+      QString msg = "In BundleLidarRangeConstraint::computeRange(): "
+          "m_rangeComputed must be positive\n";
+      throw IException(IException::Programmer, msg, _FILEINFO_);
+    }
+  }
+
+
+  /**
+   * Computes partial derivatives of range condition equation and adds contribution into the bundle
+   * adjustment normal equation matrices.
+   *
+   * @param normalsMatrix Bundle Adjustment normal equations matrix.
+   * @param N22 Normal equation matrix for the point.
+   * @param N12 Normal equations block between image and point.
+   * @param n1 Right hand side vector for the images and target body.
+   * @param n2 Right hand side vector for the point.
+   *
+   * @return bool success.
+   */
+  bool BundleLidarRangeConstraint::applyConstraint(SparseBlockMatrix &normalsMatrix,
+                                                   LinearAlgebra::MatrixUpperTriangular& N22,
+                                                   SparseBlockColumnMatrix& N12,
+                                                   LinearAlgebra::VectorCompressed& n1,
+                                                   LinearAlgebra::Vector& n2) {
+
+    if (m_simultaneousMeasure->isRejected()) {
+      return false;
+    }
+
+    // establish camera model for simultaneous measure
+    m_simultaneousMeasure->setImage();
+
+    // time of current location of simultaneous measure
+    double scaledTime = m_simultaneousMeasure->camera()->instrumentPosition()->scaledTime();
+
+    // current body fixed XYZ coordinates of lidar control point
+    SurfacePoint adjustedSurfacePoint = m_lidarControlPoint->GetAdjustedSurfacePoint();
+
+    static LinearAlgebra::Matrix coeff_range_image;          //!< Partials w/r to image position
+    static LinearAlgebra::Matrix coeff_range_point3D(1,3);   //!< Partials w/r to lidar point
+    static LinearAlgebra::Vector coeff_range_RHS(1);         //!< Right hand side of normals
+
+    // index into normal equations for this measure
+    int positionBlockIndex = m_simultaneousMeasure->positionNormalsBlockIndex();
+
+    // resize coeff_range_image matrix if necessary
+    IsisBundleObservationQsp isisObservation = m_bundleObservation.dynamicCast<IsisBundleObservation>();
+    if (!isisObservation) {
+      QString msg = "Failed to cast BundleObservation to IsisBundleObservation when applying "
+                    "lidar constraint (Point Id: "
+                    + m_simultaneousMeasure->parentControlPoint()->id() + ", Measure Serial:"
+                    + m_simultaneousMeasure->cubeSerialNumber() + ").\n";
```

Comment:
What might this suggest? I am not sure if that is information you would want to put here (if it could be too many things)

Reply:
This is 100% defensive against future changes. There's no reason for the dynamic cast above to fail right now because we check in the constructor that our image has an ISIS camera model. If the code base changes, though, and this somehow fails we'd segfault because dynamic cast failures result in a null pointer. So, this is just a check to make sure we don't attempt to dererence a null pointer. There's really no way to know why the dynamic cast failed.